pkg/util/containers/v2/metrics/collector_ecs_fargate.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// ID returns the collector ID.
+func (e *ecsFargateCollector) ID() string { return ecsFargateCollectorID }
+
+// GetContainerStats returns stats by container ID.
+func (e *ecsFargateCollector) GetContainerStats(containerID string, cacheValidity time.Duration) (*ContainerStats, error) {
+	stats, err := e.stats(containerID, cacheValidity, e.client.GetContainerStats)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertEcsStats(stats), nil
+}
+
+// GetContainerNetworkStats returns network stats by container ID.
+func (e *ecsFargateCollector) GetContainerNetworkStats(containerID string, cacheValidity time.Duration, networks map[string]string) (*ContainerNetworkStats, error) {
+	stats, err := e.stats(containerID, cacheValidity, e.client.GetContainerStats)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertEcsNetworkStats(stats.Networks, networks), nil
+}
+
+// stats returns stats by container ID, it uses an in-memory cache to reduce the number of api calls.
+// Cache expires every 2 minutes and can also be invalidated using the cacheValidity argument.
+func (e *ecsFargateCollector) stats(containerID string, cacheValidity time.Duration, clientFunc ecsStatsFunc) (*v2.ContainerStats, error) {
+	refreshRequired := e.lastScrapeTime.Add(cacheValidity).Before(time.Now())
+	cacheKey := fmt.Sprintf(statsCacheKey, containerID)
+	if cacheStats, found := cache.Cache.Get(cacheKey); found && !refreshRequired {
+		stats := cacheStats.(*v2.ContainerStats)
+		log.Debugf("Got ecs stats from cache for container %s", containerID)
+		return stats, nil
+	}
+
+	stats, err := clientFunc(context.TODO(), containerID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get container stats for %s: %w", containerID, err)
+	}
+
+	log.Debugf("Got ecs stats from ECS API for container %s", containerID)
+	e.lastScrapeTime = time.Now()
+	cache.Cache.Set(cacheKey, stats, statsCacheExpiration)
+
+	return stats, nil
+}
+
+func convertEcsStats(ecsStats *v2.ContainerStats) *ContainerStats {
+	return &ContainerStats{
+		Timestamp: time.Now(),
+		CPU:       convertCPUStats(&ecsStats.CPU),
+		Memory:    convertMemoryStats(&ecsStats.Memory),
+		IO:        convertIOStats(&ecsStats.IO),
+	}
+}
+
+func convertCPUStats(cpuStats *v2.CPUStats) *ContainerCPUStats {
+	stats := &ContainerCPUStats{}
+	convertField(&cpuStats.Usage.Total, &stats.Total)
+	convertField(&cpuStats.Usage.Kernelmode, &stats.System)
+	convertField(&cpuStats.Usage.Usermode, &stats.User)
+
+	return stats
+}
+
+func convertMemoryStats(memStats *v2.MemStats) *ContainerMemStats {
+	stats := &ContainerMemStats{}
+	convertField(&memStats.Limit, &stats.Limit)
+	convertField(&memStats.Usage, &stats.UsageTotal)
+	convertField(&memStats.Details.RSS, &stats.RSS)
+	convertField(&memStats.Details.Cache, &stats.Cache)
+
+	return stats
+}
+
+func convertIOStats(ioStats *v2.IOStats) *ContainerIOStats {
+	stats := &ContainerIOStats{}
+
+	var readBytes, writeBytes uint64
+	for _, stat := range ioStats.BytesPerDeviceAndKind {
```

Comment:
Is it possible to also feed the per device stats? I see we have major/minor in the output, can we re-use the logic used in the `system` provider?
It probably require some extra work as it probably won't work on Windows.

Reply:
in case of ecs ec2, I'd expect the system provider to be working anyway if we have access to proc fc. so I don't think it's worth implementing it in the ecs provider.

Reply:
as discussed offline, this is impossible as ecs ec2 won't be supported. marking as resolved